cdxev/build_public_bom.py

```diff
@@ -39,57 +39,83 @@ def remove_internal_information_from_properties(component: dict) -> None:
         component["properties"] = new_properties
 
 
+def clear_component(component: dict[str, Any]) -> None:
+    """
+    Removes all internal information of the component
+    and applies the same process to all sub-components
+    contained within the component.
+
+    Parameters
+    -----------
+    component: dict[str, Any]
+        A dictionary representing the component,
+        which may contain sub-components
+
+    Returns
+    -------
+    None
+    """
+    remove_internal_information_from_properties(component)
+    # The 'extract_components' function processes any nested components recursively,
+    # ensuring that all levels of sub-components are handled.
+    sub_components = extract_components(component.get("components", []))
+    for sub_component in sub_components:
+        remove_internal_information_from_properties(sub_component)
+
+
 def remove_component_tagged_internal(
-    components: Sequence[dict], path_to_schema: t.Union[Path, None]
-) -> t.Tuple[t.List[str], t.List[dict]]:
+    component: dict, validator: Draft7Validator
+) -> t.Tuple[list[str], list[dict]]:
     """
-    Removes the components marked as internal,
-    from a list of components.
-    If the component contains a properties field
-    that declares it as public, it will not be removed.
-    Internal Information will also be removed from properties fields
-    tagged with a name containing "^internal:".
+    Removes the components marked as internal (components which are valid
+    according to the schema), from a list of components. The nested
+    components will then replace the removed component if present.
+    If a component is not marked as internal, the function is
+    recursivly checking its nested components.
 
     Parameters
     ----------
     components: list[dict]
         A list of components
-    path_to_schema: Path
-        The path to a json schema defining a
-        internal component
+    validator: Draft7Validator
+        A validator to check if component is valid
+        according to the schema
 
     Returns
     -------
     list[str]
         A list with the bom-refs from the removed
         components
-    list[dict]:
-        A list of components without the property
-        "internal"
```

Comment:
This description is incorrect. It lists only the first component of the returned tuple.

Reply:
Now you're saying you're returning two things which isn't possible. What you're returning is a tuple[list[str], list[dict].
You could write, for instance: _A tuple where the first element is a list of bom-refs from the removed components and the second element is a list of the new top-level components_.